### PR TITLE
[WIP] densenet.py image segmentation support

### PR DESCRIPTION
**Pull request by ahundt**
uses Fully Convolutional Network (FCN)

### CODE REVIEW

**Comment by titu1994** [comment]
@ahundt This seems to merge my DenseNet model with my other repo for FCN DenseNet. Have you tested it to see that it correctly builds and loads the weights for normal DenseNet models?

**Comment by ahundt** [comment]
@titu1994 yes I've test it with normal densenet models

**Comment by ahundt** [comment]
I know you imagined two separate classes, but when I started making the changes I realized that the code overlap is very substantial and items like the list of layer thicknesses could be used to extend the parameterization available to the original densenet model as well, so I decided to make it so there is simply a `segmentation=True` or `segmentation=False` flag, where it defaults to the original behavior.

**Comment by titu1994** [comment]
@ahundt Cool that's pretty neat work merging the two model builder scripts. 

On another note, I see you went with Deconvolution networks instead of UpSampling 2D or SubpixelConvollution Layers like in my repo. Deconvolution layers tend to produce blocky output images, especially in domains like image inpainting and single image super resolution (I know this is for segmentation, but it FCNs can be easily applied to the above tasks as well).

Since UpSampling 2D seems to be a better at handling those blocky patches, I think an optional flag should be added to support UpSampling2D layers. Later when the SubpixelUpsampling layer is corrected to work with tensorflow, we can add that as an option as well. Reasoning is from the post http://distill.pub/2016/deconv-checkerboard/

**Comment by ahundt** [comment]
@titu1994 yeah I think there should be a parameter with a few different upsampling options, just I'd like to make those changes separate pull requests that create actual upsampling keras layers for that purpose (like the atrous link above) and add the options in those future pulls.

**Comment by titu1994** [comment]
@ahundt Oh that's useful as well. We definitely need more upsampling layers. The simple UpSampling2D is good, but dated. I'll look to see if I can get SubpixelConvolutionLayer working with TF properly and then add a PR. 

On a side note, what do the tests for upsampling layers even look like?

**Comment by ahundt** [comment]
@titu1994 we can move the upsampling conversation to #31

File: keras_contrib/applications/densenet.py [hunk 1]
```diff
@@ -301,25 +373,93 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
...
+        for block_idx in range(nb_dense_block):
+            x, nb_filter = __dense_block(x, nb_dense_block_layers[block_idx], nb_filter, growth_rate, bottleneck=bottleneck,
+                                         dropout_rate=dropout_rate, weight_decay=weight_decay)
+            # Skip connection
+            x = merge([x, skip_connection], mode='concat',
+                      concat_axis=concat_axis)
+            skip_list.append(x)
+
+            # add transition_block
+            x = transition_down_block(x, nb_filter, compression=compression, dropout_rate=dropout_rate,
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: This should be __transition_down_block

File: keras_contrib/applications/densenet.py [hunk 2]
```diff
@@ -301,25 +373,93 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
...
+                                      weight_decay=weight_decay)
+            # this is calculated inside transition_down_block
+            nb_filter = int(nb_filter * compression)
+
+            # Preserve transition for next skip connection after dense
+            skip_connection = x
+
+        # The last dense_block does not have a transition_down_block
+        x, nb_filter = dense_block(x, final_nb_layer, nb_filter, growth_rate, bottleneck=bottleneck,
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: This should be __dense_block

File: keras_contrib/applications/densenet.py [hunk 3]
```diff
@@ -14,7 +16,7 @@
 
 from keras.models import Model
 from keras.layers.core import Dense, Dropout, Activation
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Haven't imported Reshape layer

File: keras_contrib/applications/densenet.py [hunk 4]
```diff
@@ -301,25 +373,93 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
...
+
+            # Preserve transition for next skip connection after dense
+            skip_connection = x
+
+        # The last dense_block does not have a transition_down_block
+        x, nb_filter = dense_block(x, final_nb_layer, nb_filter, growth_rate, bottleneck=bottleneck,
+                                   dropout_rate=dropout_rate, weight_decay=weight_decay)
+
+        if K.image_dim_ordering() == 'th':
+            out_shape = [batch_size, nb_filter, rows // 16, cols // 16]
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Batch size is unknown in model definition. That is why I am avoiding usage of Deconvolution2D Layers. Wait for Keras 2 to arrive, in which output shape is auto inferred.

File: keras_contrib/applications/densenet.py [hunk 5]
```diff
@@ -301,25 +373,93 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
...
+        else:
+            out_shape = [batch_size, rows // 16, cols // 16, nb_filter]
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Same as above

File: keras_contrib/applications/densenet.py [hunk 6]
```diff
@@ -301,25 +373,93 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
...
+
+        # Add dense blocks and transition up block
+        for block_idx in range(nb_dense_block):
+            if K.image_dim_ordering() != 'th':
+                out_shape[3] = nb_filter
+
+            x = __transition_up_block(
+                x, nb_filters=nb_filter, type=upscaling_type, output_shape=out_shape)
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: You removed the upscaling_type parameter in my original model definition, but still call it here.

**Comment by patyork** [comment]
>How did this code even pass the tests if the code itself was riddled with bugs, I have no idea.

There aren't any tests included with this PR, so all the code had to do was pass PEP and not break anything else..

@ahundt As with everything, always run your code before sharing it, and especially before making a PR. These are similar issues as with your other PR I glanced at.

**Comment by ahundt** [comment]
How would you recommend an applications test work, anything aside from creating the model in several different configurations? I believe actual training won't be appropriate for a unit test.

**Comment by titu1994** [comment]
There's no need for a model hiding test. But ateast entire that the model's that it built are correct. 

Correctness would mean - correct number of parameters, correct loading of weights without errors, no bugs in the code. The general checks one does before making code publicly consumable by the general developers community.

**Comment by titu1994** [comment]
@ahundt If you're not in a hurry, give me till Friday. I have a test on Thursday, but once that's done, I can work on merging my two repos properly into a general framework for building DenseNets and DenseNet FCNs.

**Comment by titu1994** [comment]
@ahundt Added a PR for a working DenseNet FCN

**Comment by ahundt** [comment]
Sorry about that, I've contributed to big projects before and it isn't unusual to create a pull request to get early feedback on possible changes before everything is done and time is wasted on code or a design that could have been better, but again it should have been marked WIP and said so.

I'll close this in favor of #34 and add comments to that

**Comment by titu1994** [comment]
@ahundt Oh I didn't know about that. Generally, I don't publish any code that I haven't tested thoroughly first so it was a bit weird for me.

https://github.com/farizrahman4u/keras-contrib/pull/34 manages to merge densenet and densenet fcns neatly, albeit I am still looking into tests for SubpixelUpscaling and depth_to_scale backend functions. Other layers are pretty stable.

**Pull request closed without merging**